CMSgov/easi-app
# Change Contract Vehicle field to Contract Number - backend changes

# EASI-1977

This covers the backend changes for EASI-1977. The frontend changes will be done later and will also be merged into the `feature/EASI-1977` branch before merging everything into `main`.

## Changes and Description

- Add a `contract_number` column to the `system_intakes` database field.
- Add a `ContractNumber` field to the `SystemIntake` model, which gets fetched and updated in database storage methods.
- Contract number added to the GQL schema for _querying_ system intake contract data, resolver edited accordingly.
- Contract vehicle _replaced_ by contract number in the GQL input schema for _mutating_ system intake contract data. Resolver updated to set the old contract vehicle field to `NULL` when a contract number is supplied, to avoid confusion .
- GQL resolver tests updated and new tests added to take into account and check new functionality.
- (Unrelated to feature work) The pre-commit hook about branch names was commented out to avoid issues until https://github.com/CMSgov/easi-app/pull/1743 is merged.

## How to test this change

### Automated tests

With the backend running, run `scripts/dev test:go:only "github.com/cmsgov/easi-app/pkg/graph"`. The specific tests to look for are `TestUpdateContractDetailsImmediatelyAfterIntakeCreation`, `TestContractQueryReturnsVehicleForLegacyIntakes`, and `TestUpdateContractDetailsReplacesContractVehicleWithContractNumber`.

### Manual testing:
- Run `db:seed` to create some system intakes. The intake with ID `38e46d77-e474-4d15-a7c0-f6411221e2a4` will have no contract vehicle or contract number; the intake with ID `2ed89f9f-7fd9-4e92-89d2-cee170a44d0d` will have a contract vehicle but no number.
- Run the following GQL queries to ensure that an intake can have its contract number updated:
```
query getVehicleAndNumber {
  systemIntake(id:"38e46d77-e474-4d15-a7c0-f6411221e2a4") {
    contract {
      vehicle
      number
    }
  }
}
```
^ should return `null` for both vehicle and number.

```
mutation updateFreshIntakeWithNumber {
  updateSystemIntakeContractDetails(input:{
    id:"38e46d77-e474-4d15-a7c0-f6411221e2a4"
    contract:{
      number:"123"
    }
  }) {
    systemIntake {
      contract {
        vehicle
        number
      }
    }
  }
}
```
^ should return `"123"` for number and `null` for vehicle.

- Run the following GQL queries to make sure that a contract vehicle can be updated to a contract number:
```
query getVehicleAndNumber {
  systemIntake(id:"2ed89f9f-7fd9-4e92-89d2-cee170a44d0d") {
    contract {
      vehicle
      number
    }
  }
}
```
 ^ should return `null` for number and `"Honda"` for vehicle.

```
mutation updateExistingContractFromVehicleToNumber {
  updateSystemIntakeContractDetails(input:{
    id:"2ed89f9f-7fd9-4e92-89d2-cee170a44d0d"
    contract:{
      number:"7890"
    }
  }) {
    systemIntake {
      contract {
        vehicle
        number
      }
    }
  }
}
```
^ should return `"7890"` for number and `null` for vehicle.
